onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -216,6 +254,107 @@
+}
+
+ORT_API_STATUS_IMPL(SessionOptionsAppendExecutionProvider_DML2, _In_ OrtSessionOptions* options, OrtDmlDeviceOptions* device_opts) {
+API_IMPL_BEGIN
+    OrtDmlPerformancePreference perf_pref = device_opts->perf_pref;
+    OrtDmlDeviceFilter dev_filter = device_opts->dev_filter;
+
+    // Create DXCore Adapter Factory
+    ComPtr<IDXCoreAdapterFactory> adapterFactory;
+    ORT_THROW_IF_FAILED(::DXCoreCreateAdapterFactory(adapterFactory.GetAddressOf()));
+
+    // Get a list of all the adapters that support compute
+    ComPtr<IDXCoreAdapterList> d3D12CoreComputeAdapters;
+    GUID attributes[]{ DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE };
+    ORT_THROW_IF_FAILED(
+        adapterFactory->CreateAdapterList(_countof(attributes),
+            attributes,
+            d3D12CoreComputeAdapters.GetAddressOf()));
+
+    const uint32_t count{ d3D12CoreComputeAdapters->GetAdapterCount() };
+    int compute_only_device_id = -1;
+
+    // Struct for holding each adapter
+    enum class DeviceType { GPU, NPU, BadDevice};
+    struct AdapterInfo {
+        ComPtr<IDXCoreAdapter> Adapter;
+        DeviceType Type; // GPU or NPU
+    };
+
+    // Returns an adapter type based on the user supplied device filter
+    std::function<DeviceType (IDXCoreAdapter*)> filterAdapterTypeQuery[4] = {
+        nullptr, // enum cannot take 0 value
+        // Only GPUs are considered so all other devices aren't needed
+        [](IDXCoreAdapter* adapter){ return IsGPU(adapter) ? DeviceType::GPU : DeviceType::BadDevice; },
+        // Only NPUs are considered so all other devices aren't needed
+        [](IDXCoreAdapter* adapter){ return IsNPU(adapter) ? DeviceType::NPU : DeviceType::BadDevice; },
+        // Both GPUs and NPUs are considered
+        [](IDXCoreAdapter* adapter){ return IsGPU(adapter) ? DeviceType::GPU :
+        IsNPU(adapter) ? DeviceType::NPU : DeviceType::BadDevice; }
+    };
+
+    auto selected_adapters = std::vector<AdapterInfo>();
+    // Iterate through all compute capable adapters
+    for (uint32_t i = 0; i < count; ++i)
+    {
+        ComPtr<IDXCoreAdapter> candidateAdapter;
+        ORT_THROW_IF_FAILED(
+            d3D12CoreComputeAdapters->GetAdapter(i, candidateAdapter.GetAddressOf()));
+
+        // Add the adapters that are valid based on the device filter (GPU, NPU, or Both)
+        auto adapterType = filterAdapterTypeQuery[static_cast<int>(dev_filter)](candidateAdapter.Get());
+        if (adapterType != DeviceType::BadDevice) {
+            selected_adapters.push_back(AdapterInfo{candidateAdapter, adapterType});
+        }
+    }
+
+    // When considering both GPUs and NPUs sort them by performance preference
+    // of Default (Gpus first), HighPerformance (GPUs first), or LowPower (NPUs first)
+    if (dev_filter == OrtDmlDeviceFilter::Both) // considering both NPUs and GPUs
+    {
+        struct SortingPolicy
+        {
+            bool gpusFirst_ = true;
```

Comment:
>gpusFirst_ [](http://example.com/codeflow?start=17&length=10)

I would change the semantics of the function to npu_first instead; and I would add a comment that the default value is false because we sort gpus first in a mixed environment.